## 배열

### 배열선언




#### 리스트 생성자 사용

In [1]:
arr3 = list(range(6))
arr3

[0, 1, 2, 3, 4, 5]

#### 리스트 컨프리헨션 

In [2]:
arr4 = [x for x in range(6)]
arr4

[0, 1, 2, 3, 4, 5]

In [3]:
arr4[0] + arr4[1]  + arr4[2]  + arr4[3]  + arr4[4]  + arr4[5] 

15

### 배열 시간복잡도

- 배열 요소에 접근하는 시간복잡도 : $ O (1) $ . 소요시간이 전혀 걸리지 않음
- 배열에 맨 처음이나 중간에 삽입하는 시간복잡도 : $ O (n) $
- 배열에 맨 처음이나 중간에 삭제하는 시간복잡도 : $ O (n) $

#### 배열 선택시 고려점

- 할당할 메모리 크기 확인. 1차원 배열 최대 1000만개, 2차원 배열 최대 3000 * 3000 (900만개)
- 중간에 데이터 삽입이 많으면 시간복잡도가 높아짐

### 자주 사용하는 리스트 문법

#### append() 사용

In [4]:
my_list = [1,2,3]
my_list.append(4)

In [5]:
my_list

[1, 2, 3, 4]

#### + 연산자 사용

In [6]:
my_list = my_list + [5,6]
my_list

[1, 2, 3, 4, 5, 6]

#### insert() 함수 사용

In [7]:
my_list.insert(3, 999)
my_list

[1, 2, 3, 999, 4, 5, 6]

#### remove() 로 데이터 삭제. 값으로 삭제

In [8]:
my_list.remove(3) # 리스트에서 3을 지워

In [9]:
my_list

[1, 2, 999, 4, 5, 6]

#### 많이 사용하는 문법

In [10]:
fruits = ['apple', 'banana', 'cherry', 'apple', 'orange', 'banana', 'kiwi']

In [11]:
len(fruits)

7

In [12]:
fruits.index('banana')

1

In [13]:
fruits.index('banana', 2)

5

In [19]:
fruits.sort()

In [20]:
fruits

['apple', 'apple', 'banana', 'banana', 'cherry', 'kiwi', 'orange']

In [21]:
fruits.sort(reverse=True)

In [22]:
fruits

['orange', 'kiwi', 'cherry', 'banana', 'banana', 'apple', 'apple']

In [60]:
fruits.count('mango')

0

### 몸풀기 문제

#### 배열 정렬하기

|입력|출력|
|:--|:--|
|[1,-5,2,4,3]|[-5,1,2,3,4]|
|[2,1,1,3,2,5,4]|[1, 1, 2, 2, 3, 4, 5]|
|[6,1,7]|[1, 6, 7] |

In [23]:
def solution(arr):
    answer = []
    return answer

In [24]:
# 기본 정렬
def solution(arr):
    arr.sort()
    return arr

In [25]:
#arr1 = [1,-5,2,4,3]
arr1 = [2,1,1,3,2,5,4]
# arr1 = [6,1,7]

solution(arr1)

[1, 1, 2, 2, 3, 4, 5]

In [26]:
# 원본은 보존하면서 작업
def solution(arr):
    answer = sorted(arr, reverse=False)
    return answer

#### sort() 함수 사용않고 정렬알고리즘 구현
- $ O (n^2) $ 시간복잡도

In [27]:
import time

def bubble_sort(arr):   # 버블소트
    n = len(arr)

    for i in range(n):
        for j in range(n - i - 1):  # 한번돌고나면 맨처음 요소는 뛰어넘고 정렬 계속
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]  # 파이썬에서 temp를 만들지 않고 swap가능한 방법

    return arr

def do_sort(arr):  # sort() 사용
    arr.sort()
    return arr

def measure_time(func, arr):   # 시간측정 함수
    start_time = time.time()    # 시작시간 저장
    result = func(arr)  # 함수실행
    end_time = time.time()      # 함수 종료시간 저장

    return (end_time - start_time, result)

arr = list(range(10000))

In [33]:
arr.sort(reverse=False)

In [34]:
(bubble_time, bubble_result) = measure_time(bubble_sort, arr)

In [35]:
f'버블소트 실행 시간 : {bubble_time:.10f}'

'버블소트 실행 시간 : 1.5080649853'

In [36]:
arr.sort(reverse=True)

##### 버블소트 시간측정

In [37]:
(bubble_time, bubble_result) = measure_time(bubble_sort, arr)

In [38]:
len(bubble_result)

10000

In [39]:
f'버블소트 실행 시간 : {bubble_time:.10f}'

'버블소트 실행 시간 : 3.3834507465'

- 1~9999까지 정렬된 상태에서 버블소트 실행 1.5초 정도 소요
- 내림차순으로 역정렬된 상태에서 버블소트 실행 3.38초 정도 소요

##### 내장 소트 시간측정

In [40]:
arr = list(range(10000))
arr.sort(reverse=True)

In [41]:
sort_time, sort_result = measure_time(do_sort, arr)

In [42]:
f'내장소트 실행 시간 : {sort_time:.70f}'

'내장소트 실행 시간 : 0.0000000000000000000000000000000000000000000000000000000000000000000000'

- 내림차순으로 역정렬된 상태에서 내장소트 실행 0.00초 정도 소요

### 모의 테스트

- 공부할때 사용하는 방법 VS Code. 실제 코테 환경에서는 불가능. 프로그래머스에서 바로 구현(디버깅 불가, 코드어시스턴스 사용불가)

#### 두 개 뽑아서 더하기
- https://school.programmers.co.kr/learn/courses/30/lessons/68644

In [101]:
data = [5,0,2,7]

In [43]:
def solution(numbers):
    answer = []   # 빈 배열이 미리 생성
    # 두 수를 선택해서 모든 경우의 수를 반복문으로 구함    
    for i in range(len(numbers)):
        for j in range(i+1, len(numbers)):
            # i인덱스와 j인덱스의 값을 더해서 answer에 저장
            answer.append(numbers[i] + numbers[j])
    
    # 중복값 제거
    answer = sorted(set(answer))

    return answer

In [44]:
solution([2,1,3,4,1])

[2, 3, 4, 5, 6, 7]